subsys/usb/device/class/cdc_acm.c

```diff
@@ -624,7 +624,8 @@ static int cdc_acm_irq_tx_ready(const struct device *dev)
 {
 	struct cdc_acm_dev_data_t * const dev_data = dev->data;
 
-	if (dev_data->tx_irq_ena && dev_data->tx_ready) {
+	if (dev_data->tx_irq_ena && dev_data->tx_ready && dev_data->configured &&
+	    !dev_data->suspended) {
```

Comment:
This is touching upon a much bigger issue that never was properly addressed in Zephyr. The problem has been happening multiple times and the workaround was to disable remote wakeup via Kconfig so the device did not claim to support remote wakeup even if the device controller does.

One peculiar feature of USB suspend is that device being suspended doesn't mean that host is suspended. In fact, Linux feature called "autosuspend" is enabled by default. The way autosuspend works with CDC ACM is that if there is no data for host to send and no data is being received for autosuspend duration (e.g. 2 seconds) and **IF** the device supports Remote Wakeup, the device will be suspended. The perfectly fair expectation is that if the device has anything for host to read, then the device will initiate Remote Wakeup. Zephyr CDC ACM implementation doesn't and so we have multiple reports hitting exactly the same issue in various scenarios...

One might think than Remote Wakeup is not proper to do on "just some random log", but this problem is already solved outside Zephyr. The key is that host can disable the Remote Wakeup feature before suspending the device. If the device is suspended with disabled Remote Wakeup, then likely host is actually suspended and the user does not want the CDC ACM to be able to wakeup the system. What devices can or cannot wakeup the system from "full system suspend" is (host) system administration task and can be done e.g. with udev rules on Linux.

Therefore, in order to properly support suspend, the device code has to be aware of Remote Wakeup and issue Remote Wakeup if there is something to be sent. In order for this to happen the chicken-and-egg problem has to be solved where the rest of system should receive TX ready while USB device is suspended.

Reply:
Very insightful, thank you

The issue in my case lies in configured, not suspended, so I think at least for the issue at hand just adding the configured check would do. 

Still, currently cdc_acm_fifo_fill() rejects data while suspended so cdc_acm_irq_tx_ready() should reflect that. I don't feel competent to change the way the suspend is handled. A change on this part must be done by someone else in a different pull request.

(I could see cdc_acm_fifo_fill() using 
`if (!cdc_acm_irq_tx_ready())`
as the check instead of 
`if (!dev_data->configured || dev_data->suspended)` so a change to the way suspend works would automatically be updated. Also, cdc_acm_fifo_fill() currently doesn't check if tx_ready is true and it would add that check, too.)

Reply:
I also noticed cdc_acm_fifo_fill() can take negative len values which will have unexpected results and isn't checked.